controlplane/src/core/repositories/FederatedGraphRepository.ts

```diff
@@ -782,4 +788,188 @@ export class FederatedGraphRepository {
         }) as GraphApiKeyDTO,
     );
   }
+
+  public async registerClient(userId: string, federatedGraphName: string, clientName: string): Promise<string> {
+    if (!clientName) {
+      throw new Error('client name is empty');
+    }
+    const graph = await this.db.query.targets.findFirst({
+      columns: {},
+      with: {
+        federatedGraph: {
+          columns: { id: true },
+        },
+      },
+      where: and(
+        eq(schema.targets.name, federatedGraphName),
+        eq(schema.targets.organizationId, this.organizationId),
+        eq(schema.targets.type, 'federated'),
+      ),
+    });
+    if (graph === undefined) {
+      throw new Error(`could not find graph ${federatedGraphName}`);
+    }
+    const updatedAt = new Date();
+    await this.db
+      .insert(federatedGraphClients)
+      .values({
+        federatedGraphId: graph.federatedGraph.id,
+        name: clientName,
+        updatedAt,
+        createdBy: userId,
+        updatedBy: userId,
+      })
+      .onConflictDoUpdate({ target: [federatedGraphClients.federatedGraphId, federatedGraphClients.name], set: { updatedAt, updatedBy: userId } });
+
+    // To avoid depending on postgres, we do a second query to get the inserted client
+    const result = await this.db.query.federatedGraphClients.findFirst({
+      columns: { id: true },
+      where: and(
+        eq(federatedGraphClients.name, clientName),
+        eq(federatedGraphClients.federatedGraphId, graph.federatedGraph.id),
+      ),
+    });
+    return result!.id;
+  }
+
+  public async getClients(federatedGraphName: string): Promise<ClientDTO[]> {
+    const userRepo = new UserRepository(this.db);
+
+    const graph = await this.db.query.targets.findFirst({
+      columns: {},
+      with: {
+        federatedGraph: {
+          columns: { id: true },
+        },
+      },
+      where: and(
+        eq(schema.targets.name, federatedGraphName),
+        eq(schema.targets.organizationId, this.organizationId),
+        eq(schema.targets.type, 'federated'),
+      ),
+    });
+    if (graph === undefined) {
+      throw new Error(`could not find graph ${federatedGraphName}`);
+    }
+
+    const fedGraphClients = await this.db
+      .select({
+        name: federatedGraphClients.name,
+        id: federatedGraphClients.id,
+        createdAt: federatedGraphClients.createdAt,
+        createdBy: federatedGraphClients.createdBy,
+        lastUpdatedAt: federatedGraphClients.updatedAt,
+        lastUpdatedBy: federatedGraphClients.updatedBy,
+      })
+      .from(federatedGraphClients)
+      .where(eq(federatedGraphClients.federatedGraphId, graph.federatedGraph.id));
+
+    const clients: ClientDTO[] = [];
+
+    for (const c of fedGraphClients) {
+      const createdUser = await userRepo.byId(c.createdBy);
```

Comment:
Every client has one creator. You can create a query with a simple inner join to the user table top avoid making clients.length queries.